internal/driver/common/copy/copy_canvasobject.go

```diff
@@ -0,0 +1,112 @@
+package copy
+
+import (
+	"fyne.io/fyne/v2"
+	"fyne.io/fyne/v2/canvas"
+)
+
+type CopyCanvasObject interface {
+	fyne.CanvasObject
+
+	// Give a reference to the CanvasObject used as a source, should only be used for the pointer value, but not for any of its field
+	Ref() fyne.CanvasObject
+}
+
+type CopyReference struct {
+	ref fyne.CanvasObject
+}
+
+func (copy *CopyReference) Ref() fyne.CanvasObject {
+	return copy.ref
+}
+
+var _ CopyCanvasObject = (*CopyCircle)(nil)
+var _ CopyCanvasObject = (*CopyContainer)(nil)
+var _ CopyCanvasObject = (*CopyImage)(nil)
+var _ CopyCanvasObject = (*CopyLine)(nil)
+var _ CopyCanvasObject = (*CopyLinearGradient)(nil)
+var _ CopyCanvasObject = (*CopyRadialGradient)(nil)
+var _ CopyCanvasObject = (*CopyRaster)(nil)
+var _ CopyCanvasObject = (*CopyRectangle)(nil)
+var _ CopyCanvasObject = (*CopyText)(nil)
+
+type CopyCircle struct {
+	canvas.Circle
+	CopyReference
+}
+
+func NewCopyCircle(source *canvas.Circle) *CopyCircle {
+	return &CopyCircle{Circle: *source, CopyReference: CopyReference{source}}
+}
+
+type CopyContainer struct {
+	fyne.Container
+	CopyReference
+}
+
+func NewCopyContainer(source *fyne.Container) *CopyContainer {
+	return &CopyContainer{Container: *source, CopyReference: CopyReference{source}}
+}
+
+type CopyImage struct {
+	canvas.Image
+	CopyReference
+}
+
+func NewCopyImage(source *canvas.Image) *CopyImage {
+	return &CopyImage{Image: *source, CopyReference: CopyReference{source}}
+}
+
+type CopyLine struct {
+	canvas.Line
+	CopyReference
+}
+
+func NewCopyLine(source *canvas.Line) *CopyLine {
+	return &CopyLine{Line: *source, CopyReference: CopyReference{source}}
+}
+
+type CopyLinearGradient struct {
+	canvas.LinearGradient
+	CopyReference
+}
+
+func NewCopyLinearGradient(source *canvas.LinearGradient) *CopyLinearGradient {
+	return &CopyLinearGradient{LinearGradient: *source, CopyReference: CopyReference{source}}
+}
+
+type CopyRadialGradient struct {
+	canvas.RadialGradient
+	CopyReference
+}
+
+func NewCopyRadialGradient(source *canvas.RadialGradient) *CopyRadialGradient {
+	return &CopyRadialGradient{RadialGradient: *source, CopyReference: CopyReference{source}}
+}
+
+type CopyRaster struct {
+	canvas.Raster
+	CopyReference
+}
+
+func NewCopyRaster(source *canvas.Raster) *CopyRaster {
+	return &CopyRaster{Raster: *source, CopyReference: CopyReference{source}}
+}
+
+type CopyRectangle struct {
+	canvas.Rectangle
+	CopyReference
+}
+
+func NewCopyRectangle(source *canvas.Rectangle) *CopyRectangle {
+	return &CopyRectangle{Rectangle: *source, CopyReference: CopyReference{source}}
```

Comment:
`*source` is a read op and may race with other write ops triggered on the user side. This can happen regardless of whether the source rectangle is locked or not.

Reply:
Other places are similar.

Reply:
If all the write and read ops made by the user side is done in the same thread as where he calls Refresh, there is no need for any additional lock on his side (case with data binding for example). If the read/write ops and Refresh are happening in different thread, it is the duty of the user of the API to add lock around his read/write ops and the Refresh call.

Reply:
> If all the write and read ops made by the user side is done in the same thread as where he calls Refresh, there is no need for any additional lock on his side (case with data binding for example)

I can see this is how JavaScript hides all concurrency issues as it is single-threaded. Requiring users to know which goroutine they are (in Go) is a bit strong, IMO.

Sure we can solve all kinds of these problems if we guarantee everything to happen on the main thread, but that's strongly against Go's concurrency nature and very difficult to be realistic.

> If the read/write ops and Refresh are happening in different thread, it is the duty of the user of the API to add lock around his read/write ops and the Refresh call.

No, if Refresh and other read/write ops happen on a different thread, regardless of what the user will do, there will always be a race if Refresh internally does not do the same lock as the user might do.

Reply:
> No, if Refresh and other read/write ops happen on a different thread, regardless of what the user will do, there will always be a race if Refresh internally does not do the same lock as the user might do.

The following code would be thread safe with the code in this PR:
```
go func(){
  mtx.Lock()
  defer mtx.Unlock()
  button.Text = "test"
  button.Refresh()
}()
go func(){
  mtx.Lock()
  defer mtx.Unlock()
  button.Text = "test"
  button.Refresh()
}()
```

The other constraint I am going with here is that I am trying to keep the same API as we have today as much as possible.

Reply:
I think that code cannot guarantee thread-safe and missed the part where the renderer may access the Text attribute, which is not controlled by the lock:

```
println("on render thread")

go func(){
  mtx.Lock()
  defer mtx.Unlock()
  button.Text = "test"
  button.Refresh()
}()
go func(){
  mtx.Lock()
  defer mtx.Unlock()
  button.Text = "test"
  button.Refresh()
}()

println("on render thread: ", bottom.Text) // This is racing with others regardless of others are protected by a lock
```

Reply:
> The other constraint I am going with here is that I am trying to keep the same API as we have today as much as possible.

I can totally understand the goal here, but I think I've tried this part many times before and also discussed (briefly) in my proposal documentation that it is unlikely possible.

Reply:
> ```
> println("on render thread: ", bottom.Text) // This is racing with others regardless of others are protected by a lock
> ```

The render thread is accessing a copy of the object made at the time of the Refresh call, which means that there is no race condition as it is not accessing the same data.

Reply:
Still not that right. When the render thread starts to copy the object, it will read the text field, and it is not concurrent-safe.